src/VisualStudio/Core/Test.Next/Remote/RemoteHostClientServiceFactoryTests.cs

```diff
@@ -71,6 +71,87 @@ public async Task UpdaterService()
+            NavigateToItemKind.Delegate,
+            NavigateToItemKind.Enum,
+            NavigateToItemKind.EnumItem,
+            NavigateToItemKind.Event,
+            NavigateToItemKind.Field,
+            NavigateToItemKind.Interface,
+            NavigateToItemKind.Method,
+            NavigateToItemKind.Module,
+            NavigateToItemKind.Property,
+            NavigateToItemKind.Structure);
+
+        [Fact]
+        public async Task TestStreamingServices()
+        {
+            using var workspace = CreateWorkspace();
+
+            var exportProvider = workspace.Services.SolutionServices.ExportProvider;
+            var listenerProvider = exportProvider.GetExportedValue<AsynchronousOperationListenerProvider>();
+            var globalOptions = exportProvider.GetExportedValue<IGlobalOptionService>();
+
+            var service = workspace.Services.GetRequiredService<IRemoteHostClientProvider>();
+            using var outerClient = await service.TryGetRemoteHostClientAsync(CancellationToken.None);
+
+            // add solution, change document
+            workspace.AddSolution(SolutionInfo.Create(SolutionId.CreateNewId(), VersionStamp.Default));
+            var project = workspace.AddProject("proj", LanguageNames.CSharp);
+            var document = workspace.AddDocument(project.Id, "doc.cs", SourceText.From(""));
+
+            var tasks = new List<Task>();
+
+            // Kick off a ton of work in parallel to try to shake out race conditions with pinning/syncing.
+
+            // 100 outer loops that ensure we generate the same files with the same checksums at least 10 times.
+            for (var i = 0; i < 100; i++)
+            {
+                // 100 inner loops producing variants of the file.
+                for (var j = 0; j < 100; j++)
+                {
+                    tasks.Add(Task.Run(async () => await PerformSearchesAsync(service, document, name: "Goo" + i)));
+                }
+            }
+
+            await Task.WhenAll(tasks);
+        }
+
+        private static async Task PerformSearchesAsync(IRemoteHostClientProvider service, Document document, string name)
+        {
+            // Fork the document, with 100 variants of methods called 'name' in it.
+            var forked = document.Project.Solution.WithDocumentText(document.Id, SourceText.From(CreateText(name)));
+
+            using var client = await service.TryGetRemoteHostClientAsync(CancellationToken.None);
+
+            // Search teh forked document, ensuring we find all the results we expect.
+            var stream = client.TryInvokeStreamAsync<IRemoteNavigateToSearchService, RoslynNavigateToItem>(
+                forked,
+                (service, checksum, _) => service.SearchProjectAsync(checksum, document.Project.Id, ImmutableArray<DocumentId>.Empty, name, s_kinds, CancellationToken.None),
+                CancellationToken.None);
+
+            var count = 0;
+            await foreach (var result in stream)
```

Comment:
potentially assert that this solution is pinned until the stream ends?

Reply:
sure.  though if it wasn't, requests to sync it would fail.  but the assert is fine :)

Reply:
done!